### DISCUSSION TITLE
How can I contribute?

### DISCUSSION THREAD

**Discussion by sumeshir26** [tukaan/tukaan#111] [General]
Hi @rdbende and @Moosems,
I really love this project but struggle to find anything which is my level to contribute. 
How can I help? (Nothing too hard ofc cuz im not as good as coding as you)

**Comment by Moosems** [comment]
@sumeshir26 I'm currently working on the dialogs if you'd like to help me

**Comment by rdbende** [comment]
I'll try to write a `CONTRIBUTING.md` later today, and create some "good first issue" issues to help you (and maybe other) get started

**Reply by Moosems** [reply]
Great idea

**Comment by sumeshir26** [comment]
Yhabks!

**Comment by rdbende** [comment]
Sheesh I'm in trouble now. It's hard to figure out, in what do I need help. I grew up in a family, where if I want something to be done, I had to do it myself, and so it's difficult for me to find what others can help with (and I'm maximalist and I feel that only I can do things in the way I want it to be 😏. Nonetheless I'll try to give some description about each task, so you can see what I have in my mind. Btw, sorry about that @Moosems, when I scrapped half of your tooltip code, because I had a better idea).

However I always come up with new ideas, of what I should add to tukaan, and how it should be implemented. I have dozens of notes on my phone, that I made on the train, where I have enormous amount of time, but no WiFi :), but unfortunately I don't really have time to implement them after all.
So I'm just dumping here all the notes I find, and try to write a little context to them. If one of you would like to pick up one of them, feel free to open an issue if you want to discuss it.

Let's start with the simpler ones:

- `LabeledView` widget
a.k.a labelframe. It could be really similar to the `Frame` (probably they should even be in the same file), but with a `title` attribute (a `TextProp` from `_props.py`. It should also inherit `Frame`.
- `ButtonGroup` widget
A group of buttons, similar to `RadioGroup`. The button labels and callbacks could be specified in a dict, like `items={"Text": do_something}`. The leftmost button should have the `Left.TButton` style, the rightmost `Right.TButton`, and all the others `Middle.TButton` (similar stuff when the orientation is vertical), this way it's possible to do, that only the outer buttons have round corners. (I think my forest theme has this feature, but it might be only in one of my local branches)
- `Switch` widget
Really just a checkbutton with different style applied. Should check if the `Switch` layout is available in the current theme, and use `TCheckbutton` otherwise
- Add window icon
Currently it's not possible to add any window icon. It should be a property in `wm.py` (idk what that class is called), that accepts either a `pathlib.Path` object, or a `tukaan.Icon` object, and sets the window icon. Cross-platform twist: it should be possible to set a `png` file on every system, an `ico` Windows, and `icns` on macOS
- Add button variants
`variant` option for the `Button` widget to set styles like "outlined" (`Outline.TButton` internally), "filled" (`Filled.TButton`) and "link" (`Link.TButton`)
- Add button color variants
`color` option for the `Button` widget to set styles like "primary" (`Pimary.TButton` internally), "secondary" (`Secondary.TButton`), "normal" (`TButton`) and "danger" (`Danger.TButton`)


Sorry for any typo and grammar error, I'm dead tired 😴 😩 . I will continue this list tomorrow. I have a lot more items ;)

**Reply by Moosems** [reply]
All I really need to know is how do tcp calls work? If I can figure that out I think I can do most of these. Also, I really could use some explanation on the file dialogue so I can make it to your liking. Also, can totally relate to the maximist part

**Reply by lg-nineoneninethreenine** [reply]
Same bud

**Comment by sumeshir26** [comment]
I think I might try working on the window icons, but im not sure how to allow pngs to be set on every system...

**Reply by rdbende** [reply]
It should be possible with [`wm iconphoto`](https://www.tcl.tk/man/tcl8.6/TkCmd/wm.htm#M50)

**Comment by rdbende** [comment]
@Moosems I remember I promised you a tutorial on Tcl calls, I seem to have forgotten it :(

# The Tcl interface
The `Tcl` class is a bridge between Python and Tcl. It can be used to initialize the interpreter, to run the mainloop, to load dll files to Tk, and to call Tcl commands from Python.

```py
from tukaan._tcl import Tcl
```

## Calling Tcl from Python
> There are two methods to call Tcl code in Tukaan.

The first is the `Tcl.call` method. You can pass in arbitary number of argument of any type (see `Tcl.to` below), and it will evaluate it. The first argument is always the ruturn type of the call, if you don't expect any return it should be `None`

```py
>>> Tcl.call(None, "grid", textbox_widget, *Tcl.to_tcl_args(column=3, rowspan=8))
None
>>> Tcl.call(str, textbox_widget, "get")
foobar
```

There's a utility method called `to_tcl_args`, which can be used to convert Python keyword arguments to Tcl keyword arguments (`column=3` -> `("-column", "3")`). It returns a tuple, which you have to unpack, when passing into `Tcl.call` (with the `*` syntax)

The second is the `Tcl.eval` method, which is very similar to Python's `exec` function, except in Tcl. It doesn't do any Py-to-Tcl conversion, so you can only pass in a single string, but this means that this function is also a lot faster. Similarly to `Tcl.call` the first argument defines the return type. The second argument is the script to evaluate.

```py
>>> Tcl.eval(None, "grid .app.textbox_1 -column 3 -rowspan 8")
None
>>> Tcl.eval(str, ".app.textbox_1 get")
foobar
```


## About converting things to Tcl
You might know, that in Tcl everything is a string. Therefore Tukaan needs to convert every Python object to a string, and the strings returned from Tcl back to Python objects (this is the return value I was babbling about above). However, it isn't simply done by calling the `str()` function, but finding a way, that it can be represented in Tcl. For example when a Python funciont is passed in, it creates a new Tcl procedure at C-level, which will call the Python function (if you're really interested in this part, it's [here](https://github.com/python/cpython/blob/8feb7ab77c80968a6de6079299a39b0494b1701b/Modules/_tkinter.c#L2305) in `_tkinter.c`).
An object can also be converted to Tcl, if it has a `_name` attribute. Widgets use this attribute to store thier pathName, with which they are represented in Tk (e.g. `.frame.notebook.tab_1.button`).
If an object has a `__to_tcl__` method, but no `_name`, `Tcl.to` will call it to represent the object to Tcl. This method is useful if you want to give a name to the object on the fly. For example, Tukaan adds a `__to_tcl__` method to PIL images, which converts them to Tk images and then returns the image name, because by default PIL images aren't known by the Tcl interpreter.


If someone (@Moosems?, if you understand what the hack I'm talking about here) could revise this text, and it would be useful to have it in a `For developers` section in the documentation.

**Reply by sumeshir26** [reply]
Interesting.

**Reply by Moosems** [reply]
So it automatically converts the python object into strings in the Tcl.call and Tcl.eval?

**Reply by rdbende** [reply]
Only in `Tcl.call`. To `Tcl.eval` you pass in a string.

**Reply by Moosems** [reply]
by using widget._name?

**Reply by rdbende** [reply]
Yes
